src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -130,25 +129,27 @@ public async Task NavigateToSourceGeneratedFileAsync(SourceGeneratedDocument doc
                 File.WriteAllText(temporaryFilePath, "");
             }
 
-            var openDocumentService = _serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>();
-            var hr = openDocumentService.OpenDocumentViaProject(
-                temporaryFilePath,
-                VSConstants.LOGVIEWID.TextView_guid,
-                out _,
-                out _,
-                out _,
-                out var windowFrame);
-
-            if (ErrorHandler.Succeeded(hr) && windowFrame != null)
+            return async cancellationToken =>
             {
-                windowFrame.Show();
-            }
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var openDocumentService = _serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>();
+                var hr = openDocumentService.OpenDocumentViaProject(
+                    temporaryFilePath,
+                    VSConstants.LOGVIEWID.TextView_guid,
+                    out _,
+                    out _,
+                    out _,
+                    out var windowFrame);
+
+                if (ErrorHandler.Succeeded(hr) && windowFrame != null)
+                {
+                    windowFrame.Show();
+                }
 
-            // We should have the file now, so navigate to the right span
-            if (_openFiles.TryGetValue(temporaryFilePath, out var openFile))
-            {
-                openFile.NavigateToSpan(sourceSpan, cancellationToken);
-            }
+                // We should have the file now, so navigate to the right span
+                return _openFiles.TryGetValue(temporaryFilePath, out var openFile) &&
+                    await openFile.NavigateToSpanAsync(sourceSpan, cancellationToken).ConfigureAwait(false);
```

Comment:
This seems like a reasonable argument for your navigate method being async: at some point I'd imagine that OpenDocumentViaProject could be come async, or at least we asynchronously populate the text buffer. At that point, we'd still need this navigation to happen after that.